runtime/src/main/java/org/corfudb/runtime/clients/NettyClientProtobufRouter.java

```diff
@@ -0,0 +1,360 @@
+        // If timed out, return a exceptionally completed with the timeout.
+        try {
+            connectionFuture
+                    .get(parameters.getConnectionTimeout().toMillis(), TimeUnit.MILLISECONDS);
+        } catch (InterruptedException e) {
+            throw new UnrecoverableCorfuInterruptedError(e);
+        } catch (TimeoutException te) {
+            CompletableFuture<T> f = new CompletableFuture<>();
+            f.completeExceptionally(te);
+            return f;
+        } catch (ExecutionException ee) {
+            CompletableFuture<T> f = new CompletableFuture<>();
+            f.completeExceptionally(ee.getCause());
+            return f;
+        }
+
+        // Set up the timer and context to measure request
+        final Timer roundTripMsgTimer = CorfuRuntime.getDefaultMetrics()
+                .timer(timerNameCache.get(message.getMsgType()));
+
+        final Timer.Context roundTripMsgContext = MetricsUtils
+                .getConditionalContext(roundTripMsgTimer);
+
+        // Get the next request ID
+        final long thisRequest = requestID.getAndIncrement();
+
+        // Set the base fields for this message.
+        Request.Builder reqBuilder = request.toBuilder();
+        reqBuilder.getHeaderBuilder().setClientId(parameters.getClientId()).setRequestId(thisRequest);
+        request = reqBuilder.build();
+
+        // Generate a future and put it in the completion table.
+        final CompletableFuture<T> cf = new CompletableFuture<>();
+        outstandingRequests.put(thisRequest, cf);
+
+        // Write this message out on the channel
+        ByteBuf outBuf = PooledByteBufAllocator.DEFAULT.buffer();
+        ByteBufOutputStream requestOutputStream = new ByteBufOutputStream(outBuf);
+
+        try {
+            // Mark this message as Protobuf message (temporarily)
+            requestOutputStream.writeByte(API.PROTO_CORFU_MSG_MARK);
+            request.writeTo(requestOutputStream);
+            ctx.writeAndFlush(outBuf, ctx.voidPromise());
+            log.trace("Sent one-way request message: {}", request.getHeader());
+        } catch (IOException e) {
+            log.warn("sendRequest[{}]: Exception occurred when sending request {}, caused by {}",
+                    request.getHeader().getRequestId(), request.getHeader(), e.getCause(), e);
+        } finally {
+            IOUtils.closeQuietly(requestOutputStream);
+        }
+
+        // Generate a benchmarked future to measure the underlying request
+        final CompletableFuture<T> cfBenchmarked = cf.thenApply(x -> {
+            MetricsUtils.stopConditionalContext(roundTripMsgContext);
+            return x;
+        });
+
+        // Generate a timeout future, which will complete exceptionally
+        // if the main future is not completed.
+        final CompletableFuture<T> cfTimeout =
+                CFUtils.within(cfBenchmarked, Duration.ofMillis(timeoutResponse));
+        Request finalRequest = request;
```

Comment:
Why is this needed?

Reply:
This is actually caught by the IDE that 

> Variable used in lambda expression should be final or effectively final

and this line of code is literally inside lambda body. I did some investigation in this problem and found some explanation [here](https://docs.oracle.com/javase/specs/jls/se10/html/jls-15.html#jls-15.27.2).

> The restriction to effectively final variables prohibits access to dynamically-changing local variables, whose capture would likely introduce concurrency problems.

So I change request to an effectively final local variable, not sure is there any better way to address this problem.